dts/bindings/rtc/atmel,sam0-rtc.yaml

```diff
@@ -1,19 +1,123 @@
 # Copyright (c) 2018 omSquare s.r.o.
+# Copyright (c) 2024 Gerson Fernando Budke <nandojve@gmail.com>
 # SPDX-License-Identifier: Apache-2.0
 
 description: Atmel SAM0 RTC
 
 compatible: "atmel,sam0-rtc"
 
 include:
-  - name: rtc.yaml
+  - name: rtc-device.yaml
   - name: pinctrl-device.yaml
 
 properties:
   reg:
     required: true
 
-  clock-generator:
+  gclk:
     type: int
-    description: clock generator index
     required: true
+    description: |
+      Generic Clock Generator source
+
+  cal-constant:
+    type: int
+    required: true
+    description: |
+      Define the constant used to calculate the calibration. More
+      information can be found in the datasheet of each SoC series
+      at RTC Frequency Correction topic.
+
+      Example:
+        Correction in ppm = (FREQCORR.VALUE * 1e6 ppm) / (8192 * 128)
+
+        &rtc {
+          cal-constant = <8192 * 128>;
+        };
+
+  systimer:
+    type: boolean
+    description: |
+      Selects RTC peripheral to be used as a system timer and replace
+      the ARM systick. When this option is selected the normal RTC
+      functionality is in exclusive mode and the normal RTC functions
+      will not be available.
+
+      The systimer exclusive functionality can be enabled using the
+      following devicetree entry:
+
+        &rtc {
+          status = "okay";
+          systimer;
+        };
```

Comment:
This seems to be a software choice. rather than a hardware property. Nordic chips use their rtc0 (which is a counter) instead of systick. This is done by disabling rtc0 in the devicetree, and using it from soc code.

For platforms where the user is expected to choose the clock source to use, a SoC level Kconfig like `CONFIG_SAM_USE_RTC_FOR_SYS_CLOCK` or something along with an assert ensuring the rtc is disabled in devicetree to prevent the "default driver" to be instanciated for it seems like a more appropriate option IMO

Reply:
There are two things that makes me define this option:

1- There is a systimer driver using RTC here https://github.com/zephyrproject-rtos/zephyr/blob/b1def7145fd158e7cec50761fece73b1d4835a2e/drivers/timer/sam0_rtc_timer.c#L331

2- The RTC HW has a dedicated option to create a systimer, see 
![image](https://github.com/user-attachments/assets/7bc3c5f3-e65a-4c86-b5ac-bf940ae8918d)

Currently the developer can only chose the RTC systimer driver. After this changes he can select between systimer and RTC mode-2.

What I envision is that in the future the systimer driver should be removed in favor to make the correct implementation of the Periodic Events. In this case I still think that this option is useful to conditionally enable this feature. Having a Kconfig or a devicetree systimer option will result in the same outcome. I don't see a problem having the systimer option because it do not configure software.

The use of the systimer option in this PR will select between this two drivers. This means that if developer selects the systimer option on the RTC node he can only "see" the RTC Timer driver otherwise RTC calendar.

Reply:
The fact that the RTC can be used as the base of `sys_clock` does not make it a hardware configuration, many counters can theoretically be used as a system timer, but the implementation of `sys_clock` is a software choice. The hardware does not change based on the flag, its just used for a different purpose by zephyr.

I would suggest using a chosen node to select the use of the RTC:
```
chosen {
        atmel,sys-clock = &rtc0;
};
```
Which would make zephyr choose the `sys_clock` timer driver for it, rather than the RTC device driver. The rtc device driver would get this added:
```
#define RTC_SAM_IF_NOT_SYS_CLOCK(id)					\
	IF_DISABLED(							\
		DT_SAME_NODE(						\
			DT_CHOSEN(atmel_sys_clock),			\
			DT_DRV_INST(id)					\
		),							\
		(RTC_SAM_DEVICE(id))					\
	)

DT_INST_FOREACH_STATUS_OKAY(RTC_SAM_IF_NOT_SYS_CLOCK);
```
which only includes the driver if its not pointed to by the chosen node. This is scalable to any peripheral which could be used for sys_clock, and shows clearer intent than the boolean `systimer` added to the peripheral itself.

Reply:
Just as an example: to use arm systick instead of the rtc:
```
chosen {
        atmel,sys-clock = &systick;
};
```
It keeps the choice in one place, rather than having to remove the `systimer` property from the rtc, and add it systick?

Reply:
I intend to update the way we select `rtc1` as system timer on nordic chips to use the chosen node as I proposed here, I think it is definately a better solution than "hacks" like https://github.com/zephyrproject-rtos/zephyr/blob/c9dc58178aea7f04141d386a27c939ee1e8e1dc7/drivers/timer/nrf_rtc_timer.c#L32-L34 and we should probably take that further :)

Regarding

> See below example:
```
chosen {
        atmel,sys-clock = &rtc;
};

&rtc {
	status = "okay";
	counter-mode = "clock";
	prescaler = <1024>;
};
```
`counter-mode` should not be in the devicetree IMO, it would be set by the driver, so RTC driver would set it to "clock", sys_clock driver would set it to "counter"

Reply:
> and we should probably take that further :) I'm all for zephyr,sys-clock with appropriate helper macros to make this generic for all platforms :)

+1 but I think it is necessary an RFC for this. I can help to make this driver compatible.

> counter-mode should not be in the devicetree IMO, it would be set by the driver, so RTC driver would set it to "clock", sys_clock driver would set it to "counter"

I'm ok to drop this property at moment because only the RTC calendar will use it. The current RTC systimer I don't want touch it.

Question: How we offer these 2 options (calendar/counter) with only one driver ? Should I add later when unify both drivers ?

Reply:
I would define the implementation of sys_clock using the RTC, as a separate driver from the clock/calendar one. It is for the nordic platform atleast, which has two RTCs. One uses a sys clock driver, the other a counter api based driver :)

Reply:
I will create an RFC updating the nordic boards, I think it will be well received